### ISSUE TITLE
P0447 Introduction of std::colony to the standard library

### ISSUE THREAD

**Issue by wg21bot** [jensmaurer/papers#328]
[P0447R4](https://wg21.link/p0447r4) Introduction of std::colony to the standard library (Patrice Roy, Guy Davidson, Matthew Bentley)

**Comment by brycelelbach** [comment]
[Kona 2019-02 LEWGI Minutes](http://wiki.edg.com/bin/view/Wg21kona2019/P0447)

[P0447R6](https://wg21.link/P0447R6) `std::colony`

__Champion:__ Jonathan Wakely and Nico Josuttis

__Minute Taker:__ Vincent Reverdy

__Start Overview:__ 15:22

__Start Discussion:__ 15:59

We need a list of the things in the wild like this (prior art).

Eric F will write an implementation before Cologne.

Jonathan Wakely is thinking about putting it into GCC as an extension.

Please come back with something more wording like - we need a class synopsis.

__Start Polling:__ 16:09

__POLL: We should promise more committee time to pursuing std::colony, knowing that our time is scarce and this will leave less time for other work.__

|Strongly For|Weakly For|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|6|3|0|2|0|

__Attendance:__ 16

__End:__ 4:15

**Comment by ncm** [comment]
This is easy to implement with exactly O(1) behavior (i.e. not amortized) for iterator increment/decrement-and-dereference, with cells that may be as small as 4 bytes, and no compromises.

**Comment by brycelelbach** [comment]
Author has deferred discussion until Belfast; he hopes to have more usage experience by then.

**Comment by wg21bot** [comment]
[P0447R8](https://wg21.link/p0447r8) Introduction of std::colony to the standard library (Matt Bentley)

**Comment by wg21bot** [comment]
[P0447R9](https://wg21.link/p0447r9) Introduction of std::colony to the standard library (Matt Bentley)

**Comment by wg21bot** [comment]
[P0447R10](https://wg21.link/p0447r10) Introduction of std::colony to the standard library (Matt Bentley)

**Comment by mattreecebentley** [comment]
> 
> 
> This is easy to implement with exactly O(1) behavior (i.e. not amortized) for iterator increment/decrement-and-dereference, with cells that may be as small as 4 bytes, and no compromises.

This is in fact, not correct - the newer paper outlines why. But in brief, iterator transition between memory blocks always invokes 2 reads of the skipfield instead of one, as skipfields must be separate between blocks; with a single global skipfield you end up with vector-like reallocation when you try to remove a memory block.

**Comment by brycelelbach** [comment]
These issues are for paper tracking, not discussion of the papers.

**Comment by mattreecebentley** [comment]
> 
> 
> These issues are for paper tracking, not discussion of the papers.

Then I guess the person I was replying to should've been made aware of that too...

**Comment by wg21bot** [comment]
[P0447R11](https://wg21.link/p0447r11) Introduction of std::colony to the standard library (Matt Bentley)

**Comment by wg21bot** [comment]
[P0447R12](https://wg21.link/p0447r12) Introduction of std::colony to the standard library (Matt Bentley)

**Comment by inbal2l** [comment]
Reviewd on LEWG mailing list on 2021-02-01. 
R13 is under work.

Attaching a summary of the responses (some may already be fixed on the WIP version):

1. Recommendation to be careful not to over-specify the class, and leave room for future changes in implementation.
Author: This was considered, see also: "Is there active research in this problem space? Is it likely to change in future?" appendices. 
2. There was a discussion regarding having `colony` specified at plf:: instead of std:: to allow performance improvements which require ABI-breakage, the other side is that colony is usable by a large audience.
3. Proposal suggests to reserve `std::limits`, consider reserving a smaller scope (possibly as a member in colony) or using `colony_limits`.
Author: This was discussed broadly on SG14, making it a member creates ugly syntax, `colony_limits` is my prefered option. 
4. Consider adding constexpr implementation.
5. Elaborate on the purpose of parameter "Skipfield".
Author: Ment to allow changes for this template instance which prioritize lowered memory usage over performance. Use of this type by an implementation is not guaranteed, may change this to `prioritize_performance` / `prioritize_memory_use` enum.
6. Consider making (==, !=, <<, >>) hidden friends (need to be in the class synopsis)  / member functions (considering C++20 comparison rewrite rules) . 
(reference: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1601r0.pdf)
Note by Jonathan Wakely: We have a proposal to support comparisons between containers with different types. I will update it for <=> comparisons. For now assume you don't need to overload for different priorities.
7. Consider specifying `op==` in terms of the `std::equal`
Author: None of the other containers appear to do this, but I'm happy to do so if there's a consensus.
8. A broad discussion regarding: `advance`/`next`/`prev`, result is they are specified as hidden friends of the iterator classes.
9. Consider removing member sort()
Author: Explanation under Design Decisions->Additional notes on specific functions->void sort(). To date there's been no interest in standardising the sort algorithm which colony reference implementation uses as a non-member.
10. Consider allowing different iter types in sort() and to constrain those types with an appropriate requires clause (to allow sentinel)
Author: I think it's probably reasonable to expect that a new container should match functionality with the old ones in order to be a drop-in replacement in older code.
11. Remove duplication of paragraph which is included in memory()
12. Bike-shedding is welcome for `reshape()`

**Comment by wg21bot** [comment]
[P0447R13](https://wg21.link/p0447r13) Introduction of std::colony to the standard library (Matt Bentley)

**Comment by wg21bot** [comment]
[P0447R14](https://wg21.link/p0447r14) Introduction of std::colony to the standard library (Matt Bentley)

**Comment by wg21bot** [comment]
[P0447R15](https://wg21.link/p0447r15) Introduction of std::hive to the standard library (Matt Bentley)

**Comment by wg21bot** [comment]
[P0447R16](https://wg21.link/p0447r16) Introduction of std::hive to the standard library (Matt Bentley)

**Comment by wg21bot** [comment]
[P0447R17](https://wg21.link/p0447r17) Introduction of std::hive to the standard library (Matt Bentley)

**Comment by wg21bot** [comment]
[P0447R18](https://wg21.link/p0447r18) Introduction of std::hive to the standard library (Matt Bentley)

**Comment by wg21bot** [comment]
[P0447R19](https://wg21.link/p0447r19) Introduction of std::hive to the standard library (Matt Bentley)

**Comment by ben-craig** [comment]
## P0447R19: Introduction of std::hive to the standard library

[P0447R19](https://wg21.link/P0447R19): Introduction of std::hive to the standard library

[2022-03-22 Library Evolution Telecon Minutes](http://wiki.edg.com/bin/view/Wg21telecons2022/P0447#Library-Evolution-2022-03-22)

__Chair:__ Ben Craig

__Champion:__ Matthew Bentley 

__Minute Taker:__ Steve Downey

__SUMMARY:__

__POLL: remove operator== and operator!= from std::hive__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|9|6|1|2|0|

__Attendance:__ 22

__# of Authors:__ 1

__Author Position:__ WA

__Outcome:__ Consensus in favor of removing operator==.


__POLL: rename reshape() to change_block_capacity_limits()__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|0|3|6|3|1|

__Attendance:__ 22

__# of Authors:__ 1

__Author Position:__ N

__Outcome:__ No consensus for change


__OUTCOME:__

More review (and revision) needed.  LEWG should pick up at: 22.3.14.2 hive constructors, copy, and assignment [hive.cons] after discussing the changes provided in the revisions after R19.

**Comment by brycelelbach** [comment]
## 2022-04-05 Library Evolution Telecon 

[P0447R19](https://wg21.link/P0447R19): Introduction of std::hive to the standard library

[2022-04-05 Library Evolution Telecon Minutes](http://wiki.edg.com/bin/view/Wg21telecons2022/P0447#Library-Evolution-2022-04-05)

__Chair:__ Ben Craig

__Champion:__ Matthew Bentley 

__Minute Taker:__ Steve Downey

### Summary 

TODO: @ben-craig 

### Outcome

TODO: @ben-craig

**Comment by brycelelbach** [comment]
## 2022-04-12 Library Evolution Telecon 

[P0447R19](https://wg21.link/P0447R19): `std::hive`

[2022-04-12 Library Evolution Telecon Minutes](http://wiki.edg.com/bin/view/Wg21telecons2022/P0447#Library-Evolution-2022-04-12)

__Chair:__ Ben Craig

__Champion:__ Matt Bentley 

__Minute Taker:__ Gašper Ažman

### Summary 

TODO: @ben-craig 

### Outcome

TODO: @ben-craig

**Comment by ben-craig** [comment]
## D0447R20: std::hive

[D0447R20](https://wg21.link/P0447R20): std::hive

[2022-05-17 Library Evolution Telecon Minutes](http://wiki.edg.com/bin/view/Wg21telecons2022/P0447#Library-Evolution-2022-05-17)

__Chair:__ Ben Craig

__Minute Taker:__ Steve Downey

__Champion:__ Matthew Bentley

### Summary
Discussed get_iterator, and reviewed complexity specifications.

### Next Steps
Come back with revised paper addressing feedback

**Comment by wg21bot** [comment]
[P0447R20](https://wg21.link/p0447r20) Introduction of std::hive to the standard library (Matt Bentley)

**Comment by brycelelbach** [comment]
Deferred for now, the author's not available right now.

**Comment by ben-craig** [comment]
## P0447R21: Introduction of std::hive to the standard library

[P0447R21](https://wg21.link/P0447R21): Introduction of std::hive to the standard library

[2023-02-08 Library Evolution Telecon Minutes](http://wiki.edg.com/bin/view/Wg21telecons2022/P0447#Library-Evolution-2023-02-08)

__Champion:__ Guy Davidson

__Chair:__ Ben Craig

__Minute Taker:__ Steve Downey

__Start:__ 2023-02-08 15:45 TIMEZONE

Does this paper have:
   * Examples? Yes
   * Field experience? Yes
      * Implementation experience? Yes
      * Usage experience? Yes
      * Deployment experience? Yes
   * Performance considerations? Yes
   * Discussion of prior art? Yes
   * Changes Library Evolution previously requested? Yes
   * Wording? Yes
   * Breaking changes? No
   * Feature test macro? Yes
   * Freestanding? No
   * Format and/or iostream support? No

__End:__ 17:10

### Summary

We started with covering the revisions made to the paper.  We then discussed P2596, which proposes modifications to std::hive.  We looked at that wording, then heard a rebuttal of P2596.  There were continued questions about whether we should be spending time on std::hive.

While there is field usage and deployment experience, there were concerns that this paper isn't providing enough evidence of existing practice in the field.

__POLL:__ We like P2596 "Improve std::hive::reshape" enough to go through the wording

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|4|11|4|2|2|

__Attendance:__ 20 local, 12 remote

__# of Authors:__ 1

__Author Position:__ SF

__Outcome:__ consensus

A: I think reshape is useful, but I don't have concrete arguments
SA: Not convinced we should spend more time on this at all

__POLL:__ Spend more committee time on std::hive

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|8|4|5|2|3|

__Attendance:__ 22 local, 11 remote

__# of Authors:__ 2

__Author Position:__ 2xSF?

__Outcome:__ Not enough enough consensus to stop



__POLL:__ Apply P2596 (Improve std::hive::reshape) to P0447R21 (Introduction of std::hive to the standard library) and return to LEWG

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|2|3|8|4|3|

__Attendance:__ 27 local, 10 remote

__# of Authors:__

__Author Position:__

__Outcome:__ No consensus to change.  1 meeting for rebuttal to the rebuttal


### Next Steps

There wasn't consensus to change the hive paper.  I encourage the authors to work together to resolve the conflicts.  The author of P2596 has one meeting to provide a rebuttal to the rebuttal before we attempt to move forward with P0447.

**Comment by wg21bot** [comment]
[P0447R21](https://wg21.link/p0447r21) Introduction of std::hive to the standard library (Matt Bentley)

**Comment by mattreecebentley** [comment]
> Should be discussed with [P2857](https://wg21.link/P2857)

Bad link- which paper do you mean.

**Comment by ben-craig** [comment]
> > Should be discussed with [P2857](https://wg21.link/P2857)
> 
> Bad link- which paper do you mean.

https://isocpp.org/files/papers/P2857R0.html

**Comment by FabioFracassi** [comment]
## YYYY-MM-DD Library Evolution Telecon

[P0447R22](https://wg21.link/P0447R22): std::hive

[2023-06-16 Library Evolution Minutes](http://wiki.edg.com/bin/view/Wg21varna/P0447#Library-Evolution-2023-06-16)

__Champion:__ Matthew Bentley

__Chair:__ Fabio Fracassi

__Minute Taker:__ Robert Leahy


__POLL:__ The wording of P0447r22 needs to be improved before we can review it.

|SF|F|N|A|SA|
|-|-|-|-|-|
|2|1|8|5|6|

__Attendance:__ 26(P) + 5(R)

__Outcome:__ No consensus

__Chair Note:__ This poll was an informative poll, inside of the review session to inform whether to continue the review. The paper still needs a revision before it is seen again.

### Summary

There was significant disagreement on how detailed the complexity requirements  of individual operations should be specified. If we over specify we will loose implementation freedom, and potentially preclude performance improvements in the future. If we under specify we loose performance portability.

There is sustained opposition over this point, and -since performance is a motivating use case for this facility- it threatens the consent for the inclusion of this feature. 

### Next Steps

A small review group has volunteered to improve the wording for the next revision. 

Users with additional motivating use-cases for the proposed API are encouraged to provide them.

We will continue reviewing the revised wording.

**Comment by wg21bot** [comment]
[P0447R23](https://wg21.link/p0447r23) Introduction of std::hive to the standard library (Matt Bentley)

**Comment by ben-craig** [comment]
## P0447R24 : Introduction of std::hive to the standard library

[P0447R24](https://wg21.link/P0447R24): Introduction of std::hive to the standard library

[2023-11-07 Library Evolution Kona Minutes](http://wiki.edg.com/bin/view/Wg21kona2023/P0447#Library-Evolution-2023-11-07)

__Champion:__ Matthew Bentley

__Chair:__ Ben Craig

__Minute Taker:__ Robert Leahy

__Start:__ 2023-11-07


### Summary

__POLL: Remove relational operators for hive iterators.  Due to ABI, we likely could not add them later__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|0|2|3|5|1|

__Attendance:__ 19+5

__# of Authors:__ 1

__Author Position:__ WA

__Outcome:__ No consensus for change
 
1 abstention

__POLL: Support negative std::distance, std::ranges::distance, std::reverse_iterator, std::move_iterator and friends, but not for user supplied adaptors, for hive iterators__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
| | | | | |

__Attendance:__ 19+5

__# of Authors:__

__Author Position:__

__Outcome:__ No objection to unanimous _dissent_.  No one wants this

### Next Steps

Change swap specification so that it swaps the size limits
Decide if splice needs to mention that it invalidates end iterators
Add constexpr to block_capacity_limits()?
Work on the exception guarantees of reshape.  Strongly consider sticking with the basic exception guarantee.

Next meeting, start at reshape, and cover revisions, but do not start over from the beginning of the wording

**Comment by inbal2l** [comment]
## P0447R24 : Introduction of std::hive to the standard library

[P0447R24](https://wg21.link/P0447R24): Introduction of std::hive to the standard library

[2023-11-07 Library Evolution Kona Minutes](http://wiki.edg.com/bin/view/Wg21kona2023/P0447#Library-Evolution-2023-11-07)

__Champion:__ Matthew Bentley

__Chair:__ Inbal Levi / Nevin Liber

__Minute Taker:__ Mark Zeren

__Start:__ 2023-11-07, First Afternoon Session


### Summary

We took a poll on continue working on std::hive:

__POLL:__ We would like to devote more LEWG  time to P0443 knowing that our time is scarce and this will leave less time for other work.: 

|SF|F|N|A|SA|
| 7 | 9 | 8 | 4 | 4 |


__Attendance:__ 28+9

__# of Authors:__ 1

__Authors’ position:__ SF

__Outcome:__ consensus to continue

### Next Steps

We had a discussion on the addition of "hive" into the standard library. 
We had a consensus in favor at 2:1 ratio.

**Comment by inbal2l** [comment]
## P0447R24 : Introduction of std::hive to the standard library

[P0447R24](https://wg21.link/P0447R24): Introduction of std::hive to the standard library

[2023-11-07 Library Evolution Kona Minutes](http://wiki.edg.com/bin/view/Wg21kona2023/P0447#Library-Evolution-2023-11-07)

__Champion:__ Matthew Bentley

__Chair:__ Ben Craig

__Minute Taker:__ Robert Leahy

__Start:__ 2023-11-07


### Summary

__POLL: Remove relational operators for hive iterators.  Due to ABI, we likely could not add them later__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|0|2|3|5|1|

__Attendance:__ 19+5

__# of Authors:__ 1

__Author Position:__ WA

__Outcome:__ No consensus for change
 
1 abstention

__POLL: Support negative std::distance, std::ranges::distance, std::reverse_iterator, std::move_iterator and friends, but not for user supplied adaptors, for hive iterators__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
| | | | | |

__Attendance:__ 19+5

__# of Authors:__

__Author Position:__

__Outcome:__ No objection to unanimous _dissent_.  No one wants this

### Next Steps

Change swap specification so that it swaps the size limits
Decide if splice needs to mention that it invalidates end iterators
Add constexpr to block_capacity_limits()?
Work on the exception guarantees of reshape.  Strongly consider sticking with the basic exception guarantee.

Next meeting, start at reshape, and cover revisions, but do not start over from the beginning of the wording

**Comment by ben-craig** [comment]
## D0447R25: Introduction of std::hive to the standard library

[D0447R25](https://wg21.link/D0447R25): Introduction of std::hive to the standard library

[2023-11-10 Library Evolution Kona Minutes](http://wiki.edg.com/bin/view/Wg21kona2023/P0447#Library-Evolution-2023-11-10)

__Champion:__ Matthew Bentley

__Chair:__ Ben Craig / Inbal Levi

__Minute Taker:__ Eddie Nolan

__Start:__ 2023-11-10

### Summary

__POLL: Remove is_active from std::hive__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|7|9|1|0|3|

__Attendance:__ 21+

__# of Authors:__ 1

__Author Position:__ SA

__Outcome:__ Consensus for change

__POLL: Add precondition to get_iterator, requiring the pointee to belong to the current hive, which would remove the sentence about returning past-the-end iterator if the item is not found__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|7|9|0|2|1|

__Attendance:__ 21+

__# of Authors:__ 1

__Author Position:__ WF

__Outcome:__ Consensus



__POLL: Remove get_iterator from std::hive__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|1|5|5|4|5|

__Attendance:__ 22+

__# of Authors:__ 1

__Author Position:__ SA

__Outcome:__ No consensus




__POLL: Modify D0447R25 (Introduction of std::hive to the standard library) by removing is_active and adding a precondition to get_iterator as described above, and then send the revised paper to LWG for C++26, to be confirmed with a Library Evolution electronic poll.__

|Strongly Favor|Weakly Favor|Neutral|Weakly Against|Strongly Against|
|-|-|-|-|-|
|12|5|3|1|0|

__Attendance:__ 26+9

__# of Authors:__ 1

__Author Position:__ SF

__Outcome:__ Consensus


### Next Steps

Produce a revision with the requested changes.  That revision will go in LEWG electronic polling.

**Comment by inbal2l** [comment]
[P0447R26](https://wg21.link/P0447R26) was added to LEWG 2023 December electronic poll paper ([P30532R0](https://wg21.link/P3053R0))
(Note: Please make sure to publish it in the Dec 2023 mailing list. Please make sure you also publish P0447R25, which is the revision which we provided feedback on in the discussion above)

**Comment by mattreecebentley** [comment]
> [P0447R26](https://wg21.link/P0447R26) was added to LEWG 2023 December electronic poll paper ([P30532R0](https://wg21.link/P3053R0)) (Note: Please make sure to publish it in the Dec 2023 mailing list. Please make sure you also publish P0447R25, which is the revision which we provided feedback on in the discussion above)

Hi Inbal,
the version we discussed was the draft version, D0447R25 as noted above, so the new published version will be P0447R25. Please adjust the poll accordingly (I can alter the numbering if there's some problem with doing so).
I will have it ready for the december mailing 👍 
Cheers,
Matt

**Comment by inbal2l** [comment]
@mattreecebentley - I would prefer having D0447R25 (which was seen on Friday) published as is, as this is what was seen on Friday (and is what the polls address) and then in the same mailing list, have P0447R26 which applies the changes requested by the Friday polls. In this way, we can keep track of what was requested and when. :) Will that be OK with you?
You can approach @nliber for the technicalities on how to submit both R25, R26 to the same ML.

**Comment by inbal2l** [comment]
@mattreecebentley - I would prefer having D0447R25 (which was seen on Friday) published as is (as it is what the polls address) and then in the same mailing list, have P0447R26 which applies the changes requested by the Friday polls. In this way, we can keep track of what was requested and when. :) Will that be OK with you?
You can approach @nliber for the technicalities of submitting both R25, R26 to the same ML.

**Comment by mattreecebentley** [comment]
Cool, sounds good- Nevin says there's no issue with any of that :)
M

**Comment by inbal2l** [comment]
(tentative) [P0447R26](https://wg21.link/P0447R26) was added to LEWG 2023 December electronic poll paper ([P3053R0](https://wg21.link/P3053R0))

**Comment by mattreecebentley** [comment]
Done.

**Comment by wg21bot** [comment]
[P0447R24](https://wg21.link/p0447r24) Introduction of std::hive to the standard library (Matt Bentley, )

**Comment by wg21bot** [comment]
[P0447R25](https://wg21.link/p0447r25) Introduction of std::hive to the standard library (Matt Bentley, )

**Comment by inbal2l** [comment]
[P0447R26](https://wg21.link/p0447r26) was forwarded to LWG on LEWG 2023 December electronic poll ([P3054R0](https://wg21.link/P3054R0))

**Comment by JeffGarland** [comment]
LWG held an initial review 2024-10-18.  Notes here. https://wiki.edg.com/bin/view/Wg21telecons2024/P0447-20241018 

Additional time will be scheduled at Poland meeting to continue.

**Comment by JeffGarland** [comment]
LWG reviewed and approved in telecom 2024-12-06

Poll: add P0447r28 to C++26?

|F|A|N|
|11|0|0

**Comment by bstamour** [comment]
Meeting minutes: https://wiki.edg.com/bin/view/Wg21telecons2024/P0447-20241206

**Comment by wg21bot** [comment]
[P0447R27](https://wg21.link/p0447r27) Introduction of std::hive to the standard library (Matt Bentley)

**Comment by wg21bot** [comment]
[P0447R28](https://wg21.link/p0447r28) Introduction of std::hive to the standard library (Matt Bentley)
